[jit] refactor tryMatchSchema

zdevito opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#26499 [jit] refactor tryMatchSchema**
* #26271 [JIT] add type refinements for isinstance checks
* #26270 [jit] refactor extra sugared values
* #26269 [jit] add dynamic isinstance
* #26268 Allow types as node attributes
* #26145 [JIT] add CondValue to unify refinements and code emission

We've changed how these functions are used over time, so I cleaned up
the header file API to match. In particular:

* tryMatchSchemas was added since the overload logic got copy/pasted
into three separate locations.
* With this change, tryMatchSchema is no longer public, as it is not needed
  outside of tryMatchSchemas
* emitBuiltinFunction no longer needs a requires argument (it was always true)

* Argument order for all the schema matching stuff now puts the 'self'
builtin override last. This is only rarely used and was inconsistent with
matchSchema

* Removed OverloadedFunctionValue and OverloadedMethodValue, this functionality was folded into FunctionValue and MethodValue to keep the number of sugared values low.

Differential Revision: [D17488297](https://our.internmc.facebook.com/intern/diff/D17488297)

eellison commented on torch/csrc/jit/script/schema_matching.cpp:
```diff
@@ -417,22 +417,84 @@ MatchedSchema matchSchema(
+    const c10::optional<NamedValue>& self,
+    bool render_errors) {
+  TORCH_INTERNAL_ASSERT(schemas.size() > 0);
+  if (schemas.size() == 1) {
```
Why on schema size of 1 are we not trying `allow_conversions` ?

zdevito replied:
matchSchema does do conversions. This is avoiding the duplicate work of first trying without conversions, which is only done when there is multiple schema in order to prioritize non-converting matches first.

eellison replied:
Maybe add comment

eellison commented on torch/csrc/jit/script/schema_matching.cpp:
```diff
@@ -417,22 +417,84 @@ MatchedSchema matchSchema(
+  // error strings. If we discover it did error, then we replay it, recording
+  // the errors.
+  if (!render_errors) {
+    return matchSchemas(
```
i don't understand this - if we have reached this point in the code, none of the schemas match, can't we just throw the error ?

zdevito replied:
We start with render_errors=false, optimistically assuming that we will succeed and not spending time doing error message formatting for an overload that won't happen. If we reach here we know there will be an error, but we have not rendered the error messages, so we replay the match to render the errors.

eellison replied:
I assume this is just to cut down on compile time. It would be nice to hide the `render_errors` argument from the api since any invocation outside of this file shouldn't need to know about this. 

Related: we ran into a compile - slow down in compiler.cpp with eagerly concatenating strings and it was solved by buffering up the concatenation with lambdas: https://github.com/pytorch/pytorch/blob/264c4380b868fd985fd91cc6a6a16995287c38af/torch/csrc/jit/script/compiler.cpp#L236

eellison commented on torch/csrc/jit/script/schema_matching.cpp:
```diff
@@ -417,22 +417,84 @@ MatchedSchema matchSchema(
+                         << "The following variants are available:\n"
+                         << prefixLine(failure_messages.str(), "  ")
+                         << "\nThe original call is";
+  throw ErrorReport(loc) << failure_messages.str();
```
unreachable code

eellison commented on torch/csrc/jit/script/schema_matching.cpp:
```diff
@@ -580,11 +577,16 @@ Value* emitBuiltinCall(
+    return emitBuiltinNode(matched.second, loc, graph, name);
+  } else {
+    Function* fn = builtin_functions[matched.first - variants.size()];
+    return insertGraph(graph, *fn->graph(), matched.second.inputs).at(0);
```
keep comment around ? 
> we inline builtin calls because they are normally very small  // wrappers and are not useful for keeping around to debug

eellison replied:
Context: i ran into this when trying to add a builtin_functions with a loop that failed and the comment was helpful

eellison reviewed (commented):
Looks great!! Pasta be gone 🚫 🍝 

I have a couple comments / questions

eellison reviewed (approved):
Looks good. I think we should probably hide the `render_errors` arg with an impl pattern or something

zdevito committed:
Update on "[jit] refactor tryMatchSchema"

[jit] refactor tryMatchSchema

We've changed how these functions are used over time, so I cleaned up
the header file API to match. In particular:

* tryMatchSchemas was added since the overload logic got copy/pasted
into three separate locations.
* With this change, tryMatchSchema is no longer public, as it is not needed
  outside of tryMatchSchemas
* emitBuiltinFunction no longer needs a requires argument (it was always true)

* Argument order for all the schema matching stuff now puts the 'self'
builtin override last. This is only rarely used and was inconsistent with
matchSchema

gh-metadata: pytorch pytorch 26499 gh/zdevito/112/head

zdevito committed:
Update on "[jit] refactor tryMatchSchema"

[jit] refactor tryMatchSchema

We've changed how these functions are used over time, so I cleaned up
the header file API to match. In particular:

* tryMatchSchemas was added since the overload logic got copy/pasted
into three separate locations.
* With this change, tryMatchSchema is no longer public, as it is not needed
  outside of tryMatchSchemas
* emitBuiltinFunction no longer needs a requires argument (it was always true)

* Argument order for all the schema matching stuff now puts the 'self'
builtin override last. This is only rarely used and was inconsistent with
matchSchema

gh-metadata: pytorch pytorch 26499 gh/zdevito/112/head

zdevito committed:
Update on "[jit] refactor tryMatchSchema"

[jit] refactor tryMatchSchema

We've changed how these functions are used over time, so I cleaned up
the header file API to match. In particular:

* tryMatchSchemas was added since the overload logic got copy/pasted
into three separate locations.
* With this change, tryMatchSchema is no longer public, as it is not needed
  outside of tryMatchSchemas
* emitBuiltinFunction no longer needs a requires argument (it was always true)

* Argument order for all the schema matching stuff now puts the 'self'
builtin override last. This is only rarely used and was inconsistent with
matchSchema

gh-metadata: pytorch pytorch 26499 gh/zdevito/112/head

zdevito committed:
Update on "[jit] refactor tryMatchSchema"

[jit] refactor tryMatchSchema

We've changed how these functions are used over time, so I cleaned up
the header file API to match. In particular:

* tryMatchSchemas was added since the overload logic got copy/pasted
into three separate locations.
* With this change, tryMatchSchema is no longer public, as it is not needed
  outside of tryMatchSchemas
* emitBuiltinFunction no longer needs a requires argument (it was always true)

* Argument order for all the schema matching stuff now puts the 'self'
builtin override last. This is only rarely used and was inconsistent with
matchSchema

gh-metadata: pytorch pytorch 26499 gh/zdevito/112/head

zdevito committed:
Update on "[jit] refactor tryMatchSchema"

[jit] refactor tryMatchSchema

We've changed how these functions are used over time, so I cleaned up
the header file API to match. In particular:

* tryMatchSchemas was added since the overload logic got copy/pasted
into three separate locations.
* With this change, tryMatchSchema is no longer public, as it is not needed
  outside of tryMatchSchemas
* emitBuiltinFunction no longer needs a requires argument (it was always true)

* Argument order for all the schema matching stuff now puts the 'self'
builtin override last. This is only rarely used and was inconsistent with
matchSchema

gh-metadata: pytorch pytorch 26499 gh/zdevito/112/head

zdevito committed:
Update on "[jit] refactor tryMatchSchema"

[jit] refactor tryMatchSchema

We've changed how these functions are used over time, so I cleaned up
the header file API to match. In particular:

* tryMatchSchemas was added since the overload logic got copy/pasted
into three separate locations.
* With this change, tryMatchSchema is no longer public, as it is not needed
  outside of tryMatchSchemas
* emitBuiltinFunction no longer needs a requires argument (it was always true)

* Argument order for all the schema matching stuff now puts the 'self'
builtin override last. This is only rarely used and was inconsistent with
matchSchema

gh-metadata: pytorch pytorch 26499 gh/zdevito/112/head

zdevito committed:
Update on "[jit] refactor tryMatchSchema"

[jit] refactor tryMatchSchema

We've changed how these functions are used over time, so I cleaned up
the header file API to match. In particular:

* tryMatchSchemas was added since the overload logic got copy/pasted
into three separate locations.
* With this change, tryMatchSchema is no longer public, as it is not needed
  outside of tryMatchSchemas
* emitBuiltinFunction no longer needs a requires argument (it was always true)

* Argument order for all the schema matching stuff now puts the 'self'
builtin override last. This is only rarely used and was inconsistent with
matchSchema

gh-metadata: pytorch pytorch 26499 gh/zdevito/112/head

zdevito committed:
Update on "[jit] refactor tryMatchSchema"

[jit] refactor tryMatchSchema

We've changed how these functions are used over time, so I cleaned up
the header file API to match. In particular:

* tryMatchSchemas was added since the overload logic got copy/pasted
into three separate locations.
* With this change, tryMatchSchema is no longer public, as it is not needed
  outside of tryMatchSchemas
* emitBuiltinFunction no longer needs a requires argument (it was always true)

* Argument order for all the schema matching stuff now puts the 'self'
builtin override last. This is only rarely used and was inconsistent with
matchSchema

gh-metadata: pytorch pytorch 26499 gh/zdevito/112/head

zdevito committed:
Update on "[jit] refactor tryMatchSchema"

[jit] refactor tryMatchSchema

We've changed how these functions are used over time, so I cleaned up
the header file API to match. In particular:

* tryMatchSchemas was added since the overload logic got copy/pasted
into three separate locations.
* With this change, tryMatchSchema is no longer public, as it is not needed
  outside of tryMatchSchemas
* emitBuiltinFunction no longer needs a requires argument (it was always true)

* Argument order for all the schema matching stuff now puts the 'self'
builtin override last. This is only rarely used and was inconsistent with
matchSchema

gh-metadata: pytorch pytorch 26499 gh/zdevito/112/head

zdevito committed:
Update on "[jit] refactor tryMatchSchema"

[jit] refactor tryMatchSchema

We've changed how these functions are used over time, so I cleaned up
the header file API to match. In particular:

* tryMatchSchemas was added since the overload logic got copy/pasted
into three separate locations.
* With this change, tryMatchSchema is no longer public, as it is not needed
  outside of tryMatchSchemas
* emitBuiltinFunction no longer needs a requires argument (it was always true)

* Argument order for all the schema matching stuff now puts the 'self'
builtin override last. This is only rarely used and was inconsistent with
matchSchema

gh-metadata: pytorch pytorch 26499 gh/zdevito/112/head

zdevito committed:
Update on "[jit] refactor tryMatchSchema"

[jit] refactor tryMatchSchema

We've changed how these functions are used over time, so I cleaned up
the header file API to match. In particular:

* tryMatchSchemas was added since the overload logic got copy/pasted
into three separate locations.
* With this change, tryMatchSchema is no longer public, as it is not needed
  outside of tryMatchSchemas
* emitBuiltinFunction no longer needs a requires argument (it was always true)

* Argument order for all the schema matching stuff now puts the 'self'
builtin override last. This is only rarely used and was inconsistent with
matchSchema

gh-metadata: pytorch pytorch 26499 gh/zdevito/112/head

zdevito committed:
Update on "[jit] refactor tryMatchSchema"

[jit] refactor tryMatchSchema

We've changed how these functions are used over time, so I cleaned up
the header file API to match. In particular:

* tryMatchSchemas was added since the overload logic got copy/pasted
into three separate locations.
* With this change, tryMatchSchema is no longer public, as it is not needed
  outside of tryMatchSchemas
* emitBuiltinFunction no longer needs a requires argument (it was always true)

* Argument order for all the schema matching stuff now puts the 'self'
builtin override last. This is only rarely used and was inconsistent with
matchSchema

gh-metadata: pytorch pytorch 26499 gh/zdevito/112/head

ezyang commented:
```
Oct 10 07:42:45 test_interface (__main__.TestClassType) ... terminate called after throwing an instance of 'std::runtime_error'
Oct 10 07:42:45   what():  pybind11_object_dealloc(): Tried to deallocate unregistered instance!
Oct 10 07:42:45 Traceback (most recent call last):
Oct 10 07:42:45   File "test/run_test.py", line 458, in <module>
Oct 10 07:42:45     main()
Oct 10 07:42:45   File "test/run_test.py", line 450, in main
Oct 10 07:42:45     raise RuntimeError(message)
Oct 10 07:42:45 RuntimeError: test_jit failed! Received signal: SIGIOT
```

seems to be real

ezyang commented:
https://app.circleci.com/jobs/github/pytorch/pytorch/3145541
